### PR TITLE
Add table.shuffle

### PR DESCRIPTION
**Pull request by HybridDog**
https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
Example use case: Randomly shuffle items in a chest inventory

### CODE REVIEW

File: builtin/common/misc_helpers.lua [hunk 1]
```diff
@@ -561,6 +561,21 @@ function table.insert_all(t, other)
 end
 
 
+function table.shuffle(t, from, to, rank)
+	from = from or 1
+	to = to or #t
+	local n = to - from + 1
+	rank = rank or math.random(0, math.faculty(n)-1)
```
Comment (sfan5) [inline builtin/common/misc_helpers.lua]: when shuffling a list of 40 things, `40!` would be larger than the largest possible float (3.4E+38)
this function won't work correct in that case

is there any reason you didn't just implement a "plain" shuffle?

Comment (HybridDog) [reply builtin/common/misc_helpers.lua]: When putting items in a chest in random order during mapgen, the order should be the same if the chest is generated again at the same position with the same seed. I didn't consider the rank's size.

File: doc/lua_api.txt [hunk 2]
```diff
@@ -2424,6 +2424,11 @@ Helper functions
 * `table.insert_all(table, other_table)`:
     * Appends all values in `other_table` to `table` - uses `#table + 1` to
       find new indices.
+* `table.shuffle(table, [from], [to], [random_func])`:
+    * Shuffles elements `from` to `to` in `table` in place
+    * `from` defaults to `1`
+    * `to` defaults to `#table`
+    * `random_func` defaults to `math.random`
```
Comment (raymoo) [inline doc/lua_api.txt]: I think the documentation should describe what the function parameter should do. it doesn't look like it's required to provide the full functionality of math.random

File: doc/lua_api.txt [hunk 3]
```diff
@@ -2424,6 +2424,12 @@ Helper functions
 * `table.insert_all(table, other_table)`:
     * Appends all values in `other_table` to `table` - uses `#table + 1` to
       find new indices.
+* `table.shuffle(table, [from], [to], [random_func])`:
+    * Shuffles elements `from` to `to` in `table` in place
+    * `from` defaults to `1`
+    * `to` defaults to `#table`
+    * `random_func` defaults to `math.random`. This function receives two
+      integers as arguments and should return a random integer between them.
```
Comment (raymoo) [inline doc/lua_api.txt]: I think it should specify that it's inclusive

Comment (HybridDog) [reply doc/lua_api.txt]: Done. I thought that people can infer this from how math.random, the default function, works, but it's probably inconvenient for inexperienced persons to look up the documentation.

File: doc/lua_api.txt (hunk 2 above)
Comment (raymoo) [reply doc/lua_api.txt]: You marked this resolved but you didn't make any related change or respond with why it's unnecessary.

**Comment by HybridDog** [comment]
rebased.

**Pull request merged**